File: CS/E1267/GridViewBookmarks.cs
using System;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.Utils.Menu;
using System.Windows.Forms;
using System.Drawing;

namespace DevExpress.XtraGrid.Bookmarks {
    public class GridViewBookmarks : IDisposable {
        const string SetBookmarkStr = "Set bookmark";
        const string GotoBookmarkStr = "Go to bookmark";
        const string BookmarkStr = "Bookmark ";
        const int BookmarkCount = 10;
        GridView view;
        string keyFieldName;
        BookmarkData[] bookMarks = new BookmarkData[BookmarkCount];

        public GridViewBookmarks(GridView view, string keyFieldName) {
            this.view = view;
            this.keyFieldName = keyFieldName;
            View.MouseUp += new System.Windows.Forms.MouseEventHandler(View_MouseUp);
            View.KeyDown += new System.Windows.Forms.KeyEventHandler(View_KeyDown);
            View.CustomDrawRowIndicator += new RowIndicatorCustomDrawEventHandler(View_CustomDrawRowIndicator);
        }

        public GridView View { get { return view; } }
        public string KeyFieldName { get { return keyFieldName; } }
        protected BookmarkData[] BookMarks { get { return bookMarks; } }

        protected void SetBookmark(int rowHandle, int bookmarkIndex) {
            while (View.IsGroupRow(rowHandle)) {
                rowHandle = View.GetChildRowHandle(rowHandle, 0);
            }
            if (GetBookmarkIndex(rowHandle) == bookmarkIndex) {
                BookMarks[bookmarkIndex] = null;
            } else {
                if (BookMarks[bookmarkIndex] != null) {
                    View.InvalidateRowIndicator(GetRowHandle(bookmarkIndex));
                }
                BookMarks[bookmarkIndex] = CreateBookmarkData(rowHandle);
            }
            View.InvalidateRowIndicator(rowHandle);
        }
        protected void GotoBookmark(int bookmarkIndex) {
            if (BookMarks[bookmarkIndex] == null) return;
            View.FocusedRowHandle = GetRowHandle(bookmarkIndex);
        }
        protected bool HasBookmark(int rowHandle) {
            return GetBookmarkIndex(rowHandle) > -1;
        }
        protected virtual void DrawBookmark(RowIndicatorCustomDrawEventArgs e) {
            using (SolidBrush brush = new SolidBrush(Color.FromArgb(40, Color.DarkBlue))) {
                RectangleF bounds = new RectangleF();
                bounds.X = e.Info.Bounds.X;
                bounds.Y = e.Info.Bounds.Y;
                bounds.Width = e.Info.Bounds.Width;
                bounds.Height = e.Info.Bounds.Height;
                e.Cache.FillRectangle(brush, bounds);
            }
        }
        protected int GetRowHandle(int bookmarkIndex) {
            return View.DataController.FindRowByValue(KeyFieldName, BookMarks[bookmarkIndex].Value);
        }
        protected int GetBookmarkIndex(int rowHandle) {
            if (!View.IsDataRow(rowHandle)) return -1;
            object value = GetKeyValue(rowHandle);
            for (int i = 0; i < BookMarks.Length; i++) {
                if (BookMarks[i] == null) continue;
                if (object.Equals(BookMarks[i].Value, value)) return i;
            }
            return -1;
        }
        protected object GetKeyValue(int rowHandle) {
            return View.GetRowCellValue(rowHandle, KeyFieldName);
        }
        protected DXPopupMenu CreateMenu() {
            DXPopupMenu menu = new DXPopupMenu();
            menu.Items.Add(CreateSetBookmarkMenu());
            menu.Items.Add(CreateGotoBookmarkMenu());
            return menu;
        }
        protected DXSubMenuItem CreateSetBookmarkMenu() {
            DXSubMenuItem menuItem = new DXSubMenuItem(SetBookmarkStr);
            EventHandler eventHandler = new EventHandler(SetBookmarkMenuItem_Click);
            for (int i = 0; i < BookmarkCount; i++) {
                DXMenuCheckItem item = new DXMenuCheckItem(BookmarkStr + i.ToString());
                item.Click += eventHandler;
                item.Tag = i;
                item.Checked = BookMarks[i] != null;
                menuItem.Items.Add(item);
            }
            return menuItem;
        }
        protected DXSubMenuItem CreateGotoBookmarkMenu() {
            DXSubMenuItem menuItem = new DXSubMenuItem(GotoBookmarkStr);
            EventHandler eventHandler = new EventHandler(GotoBookmarkMenuItem_Click);
            for (int i = 0; i < BookmarkCount; i++) {
                if (BookMarks[i] != null) {
                    DXMenuItem item = new DXMenuItem(BookmarkStr + i.ToString(), eventHandler);
                    item.Tag = i;
                    menuItem.Items.Add(item);
                }
            }
            return menuItem;
        }
        protected BookmarkData CreateBookmarkData(int rowHandle) {
            return new BookmarkData(GetKeyValue(rowHandle));
        }
        void View_CustomDrawRowIndicator(object sender, RowIndicatorCustomDrawEventArgs e) {
            if (!e.Info.IsRowIndicator) return;
            if (!HasBookmark(e.RowHandle)) return;
            e.Painter.DrawObject(e.Info); // Default drawing
            e.Handled = true;
            DrawBookmark(e);
        }
        void View_MouseUp(object sender, MouseEventArgs e) {
            if (e.Button != MouseButtons.Right) return;
            GridHitInfo hitInfo = View.CalcHitInfo(e.Location);
            if (hitInfo.HitTest != GridHitTest.RowIndicator) return;
            StandardMenuManager.Default.ShowPopupMenu(CreateMenu(), View.GridControl, e.Location);
        }
        void View_KeyDown(object sender, KeyEventArgs e) {
            if (!e.Control) return;
            if (e.KeyCode < Keys.D0 || e.KeyCode > Keys.D9) return;
            if (e.Shift) {
                SetBookmark(View.FocusedRowHandle, e.KeyCode - Keys.D0);
            } else {
                GotoBookmark(e.KeyCode - Keys.D0);
            }
        }
        void SetBookmarkMenuItem_Click(object sender, EventArgs e) {
            DXMenuItem menuItem = sender as DXMenuItem;
            SetBookmark(View.FocusedRowHandle, (int)menuItem.Tag);
        }
        void GotoBookmarkMenuItem_Click(object sender, EventArgs e) {
            DXMenuItem menuItem = sender as DXMenuItem;
            GotoBookmark((int)menuItem.Tag);
        }

        #region IDisposable Members
        public void Dispose() {
            if (View != null && !View.GridControl.IsDisposed) {
                View.MouseUp -= new System.Windows.Forms.MouseEventHandler(View_MouseUp);
                View.KeyDown -= new System.Windows.Forms.KeyEventHandler(View_KeyDown);
                View.CustomDrawRowIndicator -= new RowIndicatorCustomDrawEventHandler(View_CustomDrawRowIndicator);
                this.view = null;
            }
        }
        #endregion
        protected class BookmarkData {
            object value;
            public BookmarkData(object value) {
                this.value = value;
            }
            public object Value { get { return value; } }
        }
    }
}

File: VB/E1267/GridViewBookmarks.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.Utils.Menu
Imports System.Windows.Forms
Imports System.Drawing

Namespace DevExpress.XtraGrid.Bookmarks
	Public Class GridViewBookmarks
		Implements IDisposable
		Private Const SetBookmarkStr As String = "Set bookmark"
		Private Const GotoBookmarkStr As String = "Go to bookmark"
		Private Const BookmarkStr As String = "Bookmark "
		Private Const BookmarkCount As Integer = 10
		Private view_Renamed As GridView
		Private keyFieldName_Renamed As String
		Private bookMarks_Renamed(BookmarkCount - 1) As BookmarkData

		Public Sub New(ByVal view As GridView, ByVal keyFieldName As String)
			Me.view_Renamed = view
			Me.keyFieldName_Renamed = keyFieldName
			AddHandler Me.View.MouseUp, AddressOf View_MouseUp
			AddHandler Me.View.KeyDown, AddressOf View_KeyDown
			AddHandler Me.View.CustomDrawRowIndicator, AddressOf View_CustomDrawRowIndicator
		End Sub

		Public ReadOnly Property View() As GridView
			Get
				Return view_Renamed
			End Get
		End Property
		Public ReadOnly Property KeyFieldName() As String
			Get
				Return keyFieldName_Renamed
			End Get
		End Property
		Protected ReadOnly Property BookMarks() As BookmarkData()
			Get
				Return bookMarks_Renamed
			End Get
		End Property

		Protected Sub SetBookmark(ByVal rowHandle As Integer, ByVal bookmarkIndex As Integer)
			Do While View.IsGroupRow(rowHandle)
				rowHandle = View.GetChildRowHandle(rowHandle, 0)
			Loop
			If GetBookmarkIndex(rowHandle) = bookmarkIndex Then
				BookMarks(bookmarkIndex) = Nothing
			Else
				If BookMarks(bookmarkIndex) IsNot Nothing Then
					View.InvalidateRowIndicator(GetRowHandle(bookmarkIndex))
				End If
				BookMarks(bookmarkIndex) = CreateBookmarkData(rowHandle)
			End If
			View.InvalidateRowIndicator(rowHandle)
		End Sub
		Protected Sub GotoBookmark(ByVal bookmarkIndex As Integer)
			If BookMarks(bookmarkIndex) Is Nothing Then
				Return
			End If
			View.FocusedRowHandle = GetRowHandle(bookmarkIndex)
		End Sub
		Protected Function HasBookmark(ByVal rowHandle As Integer) As Boolean
			Return GetBookmarkIndex(rowHandle) > -1
		End Function
		Protected Overridable Sub DrawBookmark(ByVal e As RowIndicatorCustomDrawEventArgs)
			Using brush As New SolidBrush(Color.FromArgb(40, Color.DarkBlue))
				Dim bounds As New RectangleF()
				bounds.X = e.Info.Bounds.X
				bounds.Y = e.Info.Bounds.Y
				bounds.Width = e.Info.Bounds.Width
				bounds.Height = e.Info.Bounds.Height
				e.Graphics.FillRectangle(brush, bounds)
			End Using
		End Sub
		Protected Function GetRowHandle(ByVal bookmarkIndex As Integer) As Integer
			Return View.DataController.FindRowByValue(KeyFieldName, BookMarks(bookmarkIndex).Value)
		End Function
		Protected Function GetBookmarkIndex(ByVal rowHandle As Integer) As Integer
			If (Not View.IsDataRow(rowHandle)) Then
				Return -1
			End If
			Dim value As Object = GetKeyValue(rowHandle)
			For i As Integer = 0 To BookMarks.Length - 1
				If BookMarks(i) Is Nothing Then
					Continue For
				End If
				If Object.Equals(BookMarks(i).Value, value) Then
					Return i
				End If
			Next i
			Return -1
		End Function
		Protected Function GetKeyValue(ByVal rowHandle As Integer) As Object
			Return View.GetRowCellValue(rowHandle, KeyFieldName)
		End Function
		Protected Function CreateMenu() As DXPopupMenu
			Dim menu As New DXPopupMenu()
			menu.Items.Add(CreateSetBookmarkMenu())
			menu.Items.Add(CreateGotoBookmarkMenu())
			Return menu
		End Function
		Protected Function CreateSetBookmarkMenu() As DXSubMenuItem
			Dim menuItem As New DXSubMenuItem(SetBookmarkStr)
			Dim eventHandler As New EventHandler(AddressOf SetBookmarkMenuItem_Click)
			For i As Integer = 0 To BookmarkCount - 1
				Dim item As New DXMenuCheckItem(BookmarkStr & i.ToString())
				AddHandler item.Click, eventHandler
				item.Tag = i
				item.Checked = BookMarks(i) IsNot Nothing
				menuItem.Items.Add(item)
			Next i
			Return menuItem
		End Function
		Protected Function CreateGotoBookmarkMenu() As DXSubMenuItem
			Dim menuItem As New DXSubMenuItem(GotoBookmarkStr)
			Dim eventHandler As New EventHandler(AddressOf GotoBookmarkMenuItem_Click)
			For i As Integer = 0 To BookmarkCount - 1
				If BookMarks(i) IsNot Nothing Then
					Dim item As New DXMenuItem(BookmarkStr & i.ToString(), eventHandler)
					item.Tag = i
					menuItem.Items.Add(item)
				End If
			Next i
			Return menuItem
		End Function
		Protected Function CreateBookmarkData(ByVal rowHandle As Integer) As BookmarkData
			Return New BookmarkData(GetKeyValue(rowHandle))
		End Function
		Private Sub View_CustomDrawRowIndicator(ByVal sender As Object, ByVal e As RowIndicatorCustomDrawEventArgs)
			If (Not e.Info.IsRowIndicator) Then
				Return
			End If
			If (Not HasBookmark(e.RowHandle)) Then
				Return
			End If
			e.Painter.DrawObject(e.Info) ' Default drawing
			e.Handled = True
			DrawBookmark(e)
		End Sub
		Private Sub View_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
			If e.Button <> MouseButtons.Right Then
				Return
			End If
			Dim hitInfo As GridHitInfo = View.CalcHitInfo(e.Location)
			If hitInfo.HitTest <> GridHitTest.RowIndicator Then
				Return
			End If
			StandardMenuManager.Default.ShowPopupMenu(CreateMenu(), View.GridControl, e.Location)
		End Sub
		Private Sub View_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
			If (Not e.Control) Then
				Return
			End If
			If e.KeyCode < Keys.D0 OrElse e.KeyCode > Keys.D9 Then
				Return
			End If
			If e.Shift Then
				SetBookmark(View.FocusedRowHandle, e.KeyCode - Keys.D0)
			Else
				GotoBookmark(e.KeyCode - Keys.D0)
			End If
		End Sub
		Private Sub SetBookmarkMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
			Dim menuItem As DXMenuItem = TryCast(sender, DXMenuItem)
			SetBookmark(View.FocusedRowHandle, CInt(Fix(menuItem.Tag)))
		End Sub
		Private Sub GotoBookmarkMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
			Dim menuItem As DXMenuItem = TryCast(sender, DXMenuItem)
			GotoBookmark(CInt(Fix(menuItem.Tag)))
		End Sub

		#Region "IDisposable Members"
		Public Sub Dispose() Implements IDisposable.Dispose
			If View IsNot Nothing AndAlso (Not View.GridControl.IsDisposed) Then
				RemoveHandler View.MouseUp, AddressOf View_MouseUp
				RemoveHandler View.KeyDown, AddressOf View_KeyDown
				RemoveHandler View.CustomDrawRowIndicator, AddressOf View_CustomDrawRowIndicator
				Me.view_Renamed = Nothing
			End If
		End Sub
		#End Region
		Protected Class BookmarkData
			Private value_Renamed As Object
			Public Sub New(ByVal value As Object)
				Me.value_Renamed = value
			End Sub
			Public ReadOnly Property Value() As Object
				Get
					Return value_Renamed
				End Get
			End Property
		End Class
	End Class
End Namespace